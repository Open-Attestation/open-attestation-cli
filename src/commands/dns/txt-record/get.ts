import { Argv } from "yargs";
import signale, { error } from "signale";
import { getLogger } from "../../../logger";
import { DnsGetTxtRecordCommand } from "./dns-command.type";
import {
  getDnsDidRecords,
  getDocumentStoreRecords,
  OpenAttestationDnsDidRecord,
  OpenAttestationDNSTextRecord,
} from "@govtechsg/dnsprove";
import { getErrorMessage } from "../../../utils";

const { trace } = getLogger("dns:txt-record");

export const command = "get [options]";

export const describe = "Get DNS TXT record entries for a specific location";

export const builder = (yargs: Argv): Argv =>
  yargs
    .option("location", {
      description: "Domain name to look up for Issuer DNS records",
      type: "string",
      demandOption: true,
    })
    .option("networkId", {
      description: "Ethereum Network (chain ID) to filter results by",
      type: "number",
    });

export const handler = async (
  args: DnsGetTxtRecordCommand
): Promise<(OpenAttestationDNSTextRecord | OpenAttestationDnsDidRecord)[]> => {
  trace(`Args: ${JSON.stringify(args, null, 2)}`);
  try {
    const allRecords: (OpenAttestationDNSTextRecord | OpenAttestationDnsDidRecord)[] = [];

    const documentStoreRecords = await getDocumentStoreRecords(args.location);
    const filteredRecords = args.networkId
      ? documentStoreRecords.filter((record) => record.netId == String(args.networkId))
      : documentStoreRecords;
    if (filteredRecords.length > 0) {
      allRecords.push(...filteredRecords);
      signale.info("List of document store records:");
      console.table(filteredRecords);
    }
    const dnsDidRecords = await getDnsDidRecords(args.location);
    if (dnsDidRecords.length > 0) {
      allRecords.push(...dnsDidRecords);
      signale.info("List of dns-did records:");
      console.table(dnsDidRecords);
    }

    if (allRecords.length === 0) {
      signale.info("No records found.");
    }
    return allRecords;
  } catch (e) {
    error(getErrorMessage(e));
  }
  return [];
};
