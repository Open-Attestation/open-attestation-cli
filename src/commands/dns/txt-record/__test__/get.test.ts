import { handler } from "../get";

describe("get", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("should return dns-txt", async () => {
    const consoleSpy = jest.spyOn(console, "table");
    await handler({ location: "donotuse.openattestation.com" });
    // dns-txt
    expect(consoleSpy).toHaveBeenNthCalledWith(1, [
      {
        addr: "0x2f60375e8144e16Adf1979936301D8341D58C36C",
        dnssec: true,
        net: "ethereum",
        netId: "3",
        type: "openatts",
      },
    ]);
    // dns-did
    expect(consoleSpy).toHaveBeenNthCalledWith(2, [
      {
        algorithm: "dns-did",
        dnssec: true,
        publicKey: "did:ethr:0xE712878f6E8d5d4F9e87E10DA604F9cB564C9a89#controller",
        type: "openatts",
        version: "1.0",
      },
    ]);
  });
});
