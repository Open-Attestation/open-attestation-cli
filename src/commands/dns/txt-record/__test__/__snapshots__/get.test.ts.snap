// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get should return dns-txt 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "addr": "0x1550DD7B63a52356194aBBe5dfeCDe3f322FF335",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0x31376b6FB90c6C16320eE8A782E4bbbCC1a96B11",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0x5FACf3cbd08AA4511Ff38D6DE64ee01A56272bBd",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0xa1de31e1812e1111F227b44a083ADaB4429a4e96",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0x7Ae2aE35f66ACE6671CB2bfb985011b0a2d9cf72",
          "dnssec": true,
          "net": "ethereum",
          "netId": "137",
          "type": "openatts",
        },
        Object {
          "addr": "0xFF36D7105715641Ca98A62a6890F86d00bBD5C25",
          "dnssec": true,
          "net": "ethereum",
          "netId": "137",
          "type": "openatts",
        },
        Object {
          "addr": "0x48399Fb88bcD031C556F53e93F690EEC07963Af3",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0x96897ba6884543bFEe6b082777E3475633165C9F",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0x9db35C07350e9a16C828dAda37fd9c2923c75812",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0xCD2dcF53B001b3998FFF7a3A13CB74bbaB27723A",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0xdA8DBd2Aaffc995F11314c0040716E791de5aEd2",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
      ],
    ],
    Array [
      Array [
        Object {
          "algorithm": "dns-did",
          "dnssec": true,
          "publicKey": "did:ethr:0x2f45d8c2b55C935d3b9fBD41fCd9ABF790d5FA25#controller",
          "type": "openatts",
          "version": "1.0",
        },
        Object {
          "algorithm": "dns-did",
          "dnssec": true,
          "publicKey": "did:ethr:0xDC674219c4057d7795d9dA3d6AB3d301BF16355D#controller",
          "type": "openatts",
          "version": "1.0",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`get should return dns-txt 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "addr": "0x1550DD7B63a52356194aBBe5dfeCDe3f322FF335",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0x31376b6FB90c6C16320eE8A782E4bbbCC1a96B11",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0x5FACf3cbd08AA4511Ff38D6DE64ee01A56272bBd",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0xa1de31e1812e1111F227b44a083ADaB4429a4e96",
          "dnssec": true,
          "net": "ethereum",
          "netId": "1",
          "type": "openatts",
        },
        Object {
          "addr": "0x7Ae2aE35f66ACE6671CB2bfb985011b0a2d9cf72",
          "dnssec": true,
          "net": "ethereum",
          "netId": "137",
          "type": "openatts",
        },
        Object {
          "addr": "0xFF36D7105715641Ca98A62a6890F86d00bBD5C25",
          "dnssec": true,
          "net": "ethereum",
          "netId": "137",
          "type": "openatts",
        },
        Object {
          "addr": "0x48399Fb88bcD031C556F53e93F690EEC07963Af3",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0x96897ba6884543bFEe6b082777E3475633165C9F",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0x9db35C07350e9a16C828dAda37fd9c2923c75812",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0xCD2dcF53B001b3998FFF7a3A13CB74bbaB27723A",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
        Object {
          "addr": "0xdA8DBd2Aaffc995F11314c0040716E791de5aEd2",
          "dnssec": true,
          "net": "ethereum",
          "netId": "3",
          "type": "openatts",
        },
      ],
    ],
    Array [
      Array [
        Object {
          "algorithm": "dns-did",
          "dnssec": true,
          "publicKey": "did:ethr:0x2f45d8c2b55C935d3b9fBD41fCd9ABF790d5FA25#controller",
          "type": "openatts",
          "version": "1.0",
        },
        Object {
          "algorithm": "dns-did",
          "dnssec": true,
          "publicKey": "did:ethr:0xDC674219c4057d7795d9dA3d6AB3d301BF16355D#controller",
          "type": "openatts",
          "version": "1.0",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
